sdk/src/main/java/io/nitric/faas/Faas.java

```diff
@@ -20,69 +20,169 @@
- * import io.nitric.faas.Response;
+ * import io.nitric.faas.http.HttpContext;
+ * import io.nitric.faas.http.HttpHandler;
+ *
+ * public class ReadHandler implements HttpHandler {
  *
- * public class Handler implements NitricFunction {
+ *     final Documents documents;
  *
- *     &commat;Override
- *     public Response handle(Trigger trigger) {
- *         return trigger.buildResponse("Hello World");
+ *     public ReadFunction(Documents documents) {
+ *         this.documents = documents;
  *     }
  *
- *     public static void main(String... args) {
- *         Faas.start(new Handler());
+ *     public HttpContext handle(HttpContext context) {
+ *         var paths = context.getRequest().getPath().split("/");
+ *         var id = paths[paths.length - 1];
+ *
+ *         try {
+ *             var json = documents.collection("examples")
+ *                 .doc(id)
+ *                 .getJson();
+ *
+ *             context.getResponse()
+ *                 .addHeader("Content-Type", "application/json")
+ *                 .data(json);
+ *
+ *         } catch (NotFoundException nfe) {
+ *             context.getResponse()
+ *                 .status(404)
+ *                 .data("Document not found: " + id);
+ *         }
+ *
+ *         return context;
+ *     }
+ *
+ *     public static void main(String[] args) {
+ *         var documents = new Documents();
+ *         var handler = new ReadHandler(documents);
+ *
+ *         new Faas().http(handler).start();
  *     }
- * }
- * </code></pre>
+ * } </code></pre>
  *
- * @see NitricFunction
+ * @see EventHandler
+ * @see EventMiddleware
+ * @see HttpHandler
+ * @see HttpMiddleware
  */
 public class Faas {
 
     private static final Logger LOGGER = Logger.getLogger("Faas");
 
-    private FaasServiceGrpc.FaasServiceStub stub = null;
+    FaasServiceGrpc.FaasServiceStub stub = null;
+    TriggerProcessor triggerProcessor = new TriggerProcessor();
+    List<EventMiddleware> eventMiddlewares = new ArrayList<>();
+    List<HttpMiddleware> httpMiddlewares = new ArrayList<>();
 
     // Public Methods -------------------------------------------------------------------
 
+    /**
+     * Register a EventHandler function.
+     *
+     * @param eventHandler the EventHandler to register (required)
+     * @return this chainable Faas object
+     */
+    public Faas event(EventHandler eventHandler) {
+        Contracts.requireNonNull(eventHandler, "eventHandler");
+
+        eventMiddlewares.add(new EventMiddleware.HandlerAdapter(eventHandler));
+        return this;
+    }
+
+    /**
+     * Register a HttpHandler function.
+     *
+     * @param httpHandler the HttpHandler to register (required)
+     * @return this chainable Faas object
+     */
+    public Faas http(HttpHandler httpHandler) {
+        Contracts.requireNonNull(httpHandler, "httpHandler");
+
+        httpMiddlewares.add(new HttpMiddleware.HandlerAdapter(httpHandler));
+        return this;
+    }
```

Comment:
> Have already spoken about this but leaving notes here for others :)

This is something I was tossing up in the go-sdk implementation, in the end I went with an exclusively middleware approach, so even the entrypoint handler is implemented with the same signature as regular middleware.

I feel like the separation of the methods and handler & middleware concepts breaks up the readability.

```java
new Faas()
  .addHttpMiddleware(middlewareA)
  .http(handler)
  .addHttpMiddleware(middlewareB)
  .start();
```
My very first thought reading this as a first time user is "What order are these methods executed in?"

After reading the code I can see it will be middlewareA -> handler -> middlewareB. But I don't the interface makes that clear.

In the other SDKs a variadic set approach is taken, at the moment this is done with a unifying method, but I think hoisting this to the entrypoint level would probably make this cleaner and easier to follow.

```java
new Faas()
  .http(
    new JWTAuth(),
    // Apply transaction here (Rollback will break the middleware chain)
    new Transaction(new Apply(), new Rollback()),
    new LogTransaction("user", "created")
  ).start()
```
> The main concern with this is familiarity for Java devs

Reply:
@medgar-nitric from our chat:

Additive method with Handler/Middleware overloads
```java
new Faas()
  .http(new JWTAuth())
  .http(new Transaction(new Apply(), new Rollback()))
  .http(new LogTransaction("user", "created"))
  .start()
```

Reply:
I think the additive method with parameter (handler/middle) overloads is the most natural from Java perspective. I like this, it is really simple to read and comprehend.

Reply:
One other option could be to support both experiences as well, not saying we have to do that, but it would actually be super easy to do, by simply making both `http` overloads optionally variadic as well

> First argument to force at least one arg, extras zero or more
`public http(HttpMiddleware m , HttpMiddlware ...extras)`
`public http(HttpHandler h , HttpHandler ...extras)`

Just a thought, it doesn't really cost anything and leaves the syntax/implementation purely up to develop preference.